File: realm-browser/src/main/java/de/jonasrottmann/realmbrowser/models/ModelsLiveData.java
package de.jonasrottmann.realmbrowser.models;

import android.annotation.SuppressLint;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;
import android.support.annotation.UiThread;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import io.realm.Realm;
import io.realm.RealmModel;

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class ModelsLiveData extends LiveData<List<ModelPojo>> {
    private final Realm realm;
    @ModelsViewModel.SortMode
    private int sortMode;
    private String filter;
    private List<ModelPojo> unfiltered;

    ModelsLiveData(Realm realm, int sortMode, @NonNull String filter) {
        this.realm = realm;
        this.sortMode = sortMode;
        this.filter = filter;
        loadModels();
    }

    @NonNull
    private static List<ModelPojo> sortPojos(@NonNull List<ModelPojo> pojos, @ModelsViewModel.SortMode int sortMode) {
        Collections.sort(pojos, new Comparator<ModelPojo>() {
            @Override
            public int compare(ModelPojo o1, ModelPojo o2) {
                return o1.getKlass().getSimpleName().compareTo(o2.getKlass().getSimpleName());
            }
        });
        if (sortMode == ModelsViewModel.SortMode.DESC) {
            Collections.reverse(pojos);
        }
        return pojos;
    }

    @NonNull
    private static List<ModelPojo> filterPojos(@NonNull List<ModelPojo> pojos, @Nullable String filter) {
        ArrayList<ModelPojo> filteredPojos = null;
        if (filter != null && !filter.isEmpty()) {
            filteredPojos = new ArrayList<>();
            for (ModelPojo pojo : pojos) {
                if (pojo.getKlass().getSimpleName().toLowerCase().contains(filter.toLowerCase())) {
                    filteredPojos.add(pojo);
                }
            }
        }
        return filteredPojos == null ? pojos : filteredPojos;
    }

    void onSortModeChanged(@ModelsViewModel.SortMode int sortMode) {
        this.sortMode = sortMode;
        updateValueAsync();
    }

    void onFilterChanged(@Nullable String filter) {
        this.filter = filter;
        updateValueAsync();
    }

    int getSortMode() {
        return sortMode;
    }

    String getFilter() {
        return filter;
    }

    void refreshData() {
        loadModels();
    }

    @UiThread
    private void loadModels() {
        ArrayList<Class<? extends RealmModel>> realmModelClasses = new ArrayList<>(realm.getConfiguration().getRealmObjectClasses());
        ArrayList<ModelPojo> pojos = new ArrayList<>();
        for (Class<? extends RealmModel> klass : realmModelClasses) {
            pojos.add(new ModelPojo(klass, realm.where(klass).findAll().size()));
        }
        this.unfiltered = pojos;
        updateValueAsync();
    }

    @SuppressLint("StaticFieldLeak")
    private void updateValueAsync() {
        new AsyncTask<Void, Void, List<ModelPojo>>() {
            @Override
            protected List<ModelPojo> doInBackground(Void... voids) {
                return filterPojos(sortPojos(unfiltered, ModelsLiveData.this.sortMode), ModelsLiveData.this.filter);
            }

            @Override
            protected void onPostExecute(List<ModelPojo> modelsPojos) {
                setValue(modelsPojos);
            }
        }.execute();
    }
}

File: realm-browser/src/main/java/de/jonasrottmann/realmbrowser/models/ModelsLiveData.kt
package de.jonasrottmann.realmbrowser.models

import android.arch.lifecycle.LiveData
import android.support.annotation.RestrictTo
import android.support.annotation.UiThread
import android.text.TextUtils
import de.jonasrottmann.realmbrowser.ioThread
import io.realm.Realm
import java.util.*
import kotlin.collections.ArrayList

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class ModelsLiveData(private val realm: Realm, sortMode: Int, filter: String) : LiveData<List<ModelPojo>>() {

    @SortMode.SortMode
    var sortMode: Int = SortMode.ASC
        set(value) {
            field = value
            updateValueAsync()
        }
    var filter: String = ""
        set(value) {
            field = value.trim()
            updateValueAsync()
        }
    private var unfiltered: List<ModelPojo> = ArrayList()

    init {
        this.sortMode = sortMode
        this.filter = filter
        loadModels()
    }

    private fun sortPojos(pojos: List<ModelPojo>, @SortMode.SortMode sortMode: Int): List<ModelPojo> {
        Collections.sort(pojos) { o1, o2 -> o1.klass.simpleName.compareTo(o2.klass.simpleName) }
        if (sortMode == SortMode.DESC) {
            Collections.reverse(pojos)
        }
        return pojos
    }

    private fun filterPojos(pojos: List<ModelPojo>, filter: String): List<ModelPojo> {
        var filteredPojos: ArrayList<ModelPojo>? = null
        if (TextUtils.isGraphic(filter)) {
            filteredPojos = pojos.filterTo(ArrayList()) { it.klass.simpleName.toLowerCase().contains(filter.toLowerCase()) }
        }
        return if (filteredPojos == null) pojos else filteredPojos
    }

    @UiThread
    fun loadModels() {
        val realmModelClasses = ArrayList(realm.configuration.realmObjectClasses)
        val pojos = realmModelClasses.map { ModelPojo(it, realm.where(it).findAll().size.toLong()) }
        this.unfiltered = pojos
        updateValueAsync()
    }

    private fun updateValueAsync() {
        ioThread {
            postValue(filterPojos(sortPojos(unfiltered, this@ModelsLiveData.sortMode), this@ModelsLiveData.filter))
        }
    }
}
